src/util/interval_union.cpp

```diff
@@ -0,0 +1,236 @@
+/*******************************************************************\
+
+Module: Util
+
+Author: Diffblue Limited
+
+\*******************************************************************/
+
+#include "interval_union.h"
+#include "arith_tools.h"
+#include "range.h"
+#include "string_utils.h"
+#include <regex>
+#include <util/invariant.h>
+#include <util/std_expr.h>
+
+interval_uniont interval_uniont::all_integers()
+{
+  return of_interval(intervalt{});
+}
+
+interval_uniont interval_uniont::greater_or_equal(const mp_integer &value)
+{
+  interval_uniont interval_union{};
+  intervalt interval;
+  interval.make_ge_than(value);
+  interval_union.intervals.emplace_back(std::move(interval));
+  return interval_union;
+}
+
+interval_uniont interval_uniont::smaller_or_equal(const mp_integer &value)
+{
+  interval_uniont interval_union{};
+  intervalt interval;
+  interval.make_le_than(value);
+  interval_union.intervals.emplace_back(std::move(interval));
+  return interval_union;
+}
+
+interval_uniont
+interval_uniont::make_intersection(const interval_uniont &other) const
+{
+  interval_uniont result{};
+  auto it1 = intervals.begin();
+  auto it2 = other.intervals.begin();
+  while(it1 != intervals.end() && it2 != other.intervals.end())
+  {
+    intervalt intersection = *it1;
+    intersection.intersect_with(*it2);
+    if(!intersection.empty())
+      result.intervals.emplace_back(std::move(intersection));
+    if(it1->upper < it2->upper)
+      ++it1;
+    else
+      ++it2;
+  }
+  return result;
+}
+
+interval_uniont interval_uniont::make_union(const interval_uniont &other) const
+{
+  if(intervals.empty())
+    return other;
+  if(other.intervals.empty())
+    return *this;
+
+  interval_uniont result{};
+  auto it1 = intervals.begin();
+  auto it2 = other.intervals.begin();
+  if(!it1->lower_set)
+  {
+    if(!it2->lower_set)
+    {
+      result = smaller_or_equal(std::max(it1->upper, it2->upper));
+      ++it1;
+      ++it2;
+    }
+    else
+    {
+      result.intervals.push_back(*it1);
+      ++it1;
+    }
+  }
+  else
+  {
+    if(!it2->lower_set)
+    {
+      result.intervals.push_back(*it2);
+      ++it2;
+    }
+    else if(it1->lower <= it2->lower)
+    {
+      result.intervals.push_back(*it1);
+      ++it1;
+    }
+    else
+    {
+      result.intervals.push_back(*it2);
+      ++it2;
+    }
+  }
+  while(it1 != intervals.end() || it2 != other.intervals.end())
```

Comment:
Invariant here that it1 and it2 are both either `end()` or have `lower_set`?

Reply:
I guess you mean `upper_set`?

Reply:
I actually meant `lower_set`, because we just executed a prelude specifically for intervals open at the lower end -- I expected at this point any interval is closed at the bottom

Reply:
:+1: I got it now